core/src/main/scala/org/http4s/headers/Forwarded.scala

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright 2013-2020 http4s.org
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.http4s.headers
+
+import java.net.{Inet4Address, Inet6Address}
+
+import cats.data.NonEmptyList
+import org.http4s.parser.{ForwardedModelParsers, HttpHeaderParser}
+import org.http4s.util.Writer
+import org.http4s.{Header, HeaderKey, ParseResult, Uri}
+
+object Forwarded
+    extends HeaderKey.Internal[Forwarded]
+    with HeaderKey.Recurring
+    with ForwardedModelParsers {
+
+  final case class Node(nodeName: Node.Name, nodePort: Option[Node.Port] = None)
+
+  object Node {
+    def apply(nodeName: Name, nodePort: Port): Node = apply(nodeName, Some(nodePort))
+
+    sealed trait Name extends Any
+
+    object Name {
+      final case class Ipv4(address: Uri.Ipv4Address) extends AnyVal with Name
+      final case class Ipv6(address: Uri.Ipv6Address) extends AnyVal with Name
+      case object Unknown extends Name
+
+      def apply(address: Uri.Ipv4Address): Name = Name.Ipv4(address)
+      def apply(address: Inet4Address): Name = apply(Uri.Ipv4Address.fromInet4Address(address))
+      def apply(a: Byte, b: Byte, c: Byte, d: Byte): Name = apply(Uri.Ipv4Address(a, b, c, d))
+
+      def apply(address: Uri.Ipv6Address): Name = Name.Ipv6(address)
+      def apply(address: Inet6Address): Name = apply(Uri.Ipv6Address.fromInet6Address(address))
+      def apply(a: Short, b: Short, c: Short, d: Short, e: Short, f: Short, g: Short, h: Short)
+          : Name = apply(Uri.Ipv6Address(a, b, c, d, e, f, g, h))
+    }
+
+    sealed trait Port extends Any
+
+    object Port {
+      final case class Num(value: Int) extends AnyVal with Port
+
+      def apply(num: Int): Port = Port.Num(num)
+      def unapply(port: Port): Option[Int] =
+        PartialFunction.condOpt(port) {
+          case Num(num) => num
+        }
+    }
+
+    final case class Obfuscated(value: String) extends AnyVal with Name with Port
```

Comment:
Reading the spec, I expected `obfnode` to be in `Node` and `obfport` to be in `Port`.  The grammar is the same for this type, but it seems weird to me for the type to do double duty.

Reply:
I agree that it is a controversial approach. I was just thinking – what would be advantages of making them separated?
I mean it would not only require to create two different (although identical) case classes here, but also two parser rules for creating these classes. No doubt we could share some of their implementation, but it would increase overall code complexity anyway (remember, there still should be `fromString` as well as rendering implemented for both of them). On the other side, I couldn't come up with some real disadvantages besides, yeah, some weirdness.

I'm just trying to come up with some compromise solution that would let us avoid code duplication, preserve type safety and keep code complexity on some maintainable level. WDYT?

Reply:
This comment is about minutiae, so I don't feel too strongly.

`Obfuscated` isn't as untyped as `String`, but it still raises the question, "an obfuscated what?"  It's not a particularly meaningful type on its own.  Sharing makes the implementation more concise, but isn't semantically meaningful in an HTTP sense.

In the unlikely event that the next version of that RFC allows some other character in an obfuscated host, we've got room to change it.  If the two unrelated concepts share a type, we'd have to wait for the next breaking release to implement it.

I don't think either argument matters much in this corner case of a supplemental RFC.  I'm thinking more in general design.  And you're right that the repetition is annoying, so I'm content either way.